Gordon360/Hubs/SignalRHub.cs

```diff
@@ -20,30 +20,80 @@ namespace Gordon360.Hubs
     [HubName("ChatHub")]
     public class ChatHub : Hub
     {
-        //public DirectMessageService _DirectMessageService = new DirectMessageService();
+        public async Task ManageGroupAsync(string connectionId, string groupName, string function)
+        {
+            if (connectionId == null)
+            {
+                throw new ArgumentNullException(nameof(connectionId));
+            }
+
+            if (groupName == null)
+            {
+                throw new ArgumentNullException(nameof(groupName));
+            }
+            
+
+            if (function == "add")
+            {
+                await Groups.Add(connectionId, groupName);
+
+            } else
+            {
+                Groups.Remove(connectionId, groupName);
+            }
+
+        }
+
 
-        public async Task refreshMessages(List<string> connectionIds, IEnumerable<MessageViewModel> message)
+        public async Task refreshMessages(List<string> userIds, SendTextViewModel message, string userId)
          {
-            foreach(string connections in connectionIds)
+            DirectMessageService _DirectMessageService = new DirectMessageService();
+            var connectionIds =_DirectMessageService.GetUserConnectionIds(userIds);
+            
+            foreach (var connections in connectionIds)
             {
-                await Groups.Add(connections, "list");
+                foreach (var userConnections in connections)
+                {
+                    await ManageGroupAsync(userConnections.connection_id, "list", "add");
+                }
+            }
+
+            await Clients.Group("list").SendAsync(message, userId);
+
+            foreach (var connections in connectionIds)
+            {
+                foreach (var userConnections in connections)
+                {
+                     ManageGroupAsync(userConnections.connection_id, "list", "remove");
+                }
             }
-            await Clients.Group("list").SendAsync("Received message refresh, Message: ", message);
          }
 
         public string savedUserId;
         public string savedConnectionId;
 
-        public  void saveConnection(string id)
+        public async Task saveFunction(string id)
         {
             DirectMessageService _DirectMessageService = new DirectMessageService();
+            _DirectMessageService.StoreUserConnectionIds(id, Context.ConnectionId);
+        }
+
+        public  async Task saveConnection(string id)
+        {
             savedUserId = id;
             savedConnectionId = Context.ConnectionId;
-            _DirectMessageService.StoreUserConnectionIds(id, Context.ConnectionId);
-             Clients.All.BroadcastMessage("It's working boss");
+            await saveFunction(id);
+
+            Clients.All.BroadcastMessage("Saved Connection!");
 
         }
 
```

Comment:
The reasoning behind this is to create a hacked way for the service calls to be async. By itself, I cannot call a service function that implements await. But if I create a helper function I can declare it as async and call that function using await.

Reply:
I'm not sure I understand, since the service function still ends up being async, so it seems to me like it will still have to be awaited or someohow resolve synchronously when it's called. But I am inexperienced with both C# async/await and SignalR so I will defer to you on this one.